### PR TITLE
[Cache] Fix Memory leak

### LINKED ISSUE #43918: [Cache] Memory leak when using the cache
### Symfony version(s) affected

symfony/framework-bundle:  
`5.3.10`, `5.2.14`, `5.1.11`, `5.4.x-dev@24649c622776b36158f80b81b9455987b8129793` , `6.0.x-dev@5284b7f8419f9ed224d8d00d6c945d566343cd83`, `4.4.31`, and I assume every version in between.

PHP: `7.4.15`, `8.0.11`, `8.0.12`, `PHP 8.1.0RC5`

### Description

When using the Symfony cache in a certain way (e.g. calling the cache within a closure that is resolving another cache value), there is a memory leak. 

It's quite a strange bug and I don't know what is causing it. I was able to reproduce it (more info in the reproducer's readme). I'm not sure if this could be a PHP bug either.

In the reproducer the memory increase is not that high. However, in our own application the memory increase is ~10 MB per call to the method that does a bunch of caching. This is especially a problem in our message consumers.

I'll test the reproducer against some other Symfony versions too.

### How to reproduce

Reproducer: https://github.com/Jeroeny/reproduce/tree/cache

### Possible Solution

In our own application `gc_collect_cycles()` mitigated the problem. But not in the reproducer.

### PR DESCRIPTION
**Pull request by a1812**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix #43918 
| License       | MIT
| Doc PR        | 

Rough decision as an example of how to stop a leak

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
@Jeroeny can you please try this patch a report back if it fixes your issue?

**Comment by Jeroeny** [comment]
> @Jeroeny can you please try this patch a report back if it fixes your issue?

It no longer seems to leak in the reproducer after having applied these changes. 👍

File: src/Symfony/Component/Cache/Traits/AbstractAdapterTrait.php [hunk 1]
```diff
@@ -379,18 +378,14 @@ private function getId($key)
             }
         }
 
-        if (\is_string($key) && isset($this->ids[$key])) {
```
Comment (Jeroeny) [inline src/Symfony/Component/Cache/Traits/AbstractAdapterTrait.php]: Maybe the `this->ids` property can be removed entirely, as this is the only place it's read as far as I can see.

**Comment by nicolas-grekas** [comment]
This property has been added to save calling `validateKey()` and to save hashing long keys.

Since #40317, it's possible to skip calling `validateKey()` so I'm fine removing for that part.
But for hashing, this would still be useful.

This property has been added with the assumption that the number of different keys is upper bounded.
We could enforce this upper bound by limiting the size of the `ids` array instead. Eg resetting it when it reaches 1000 items?

**Comment by a1812** [comment]
> But for hashing, this would still be useful.
 > This property has been added with the assumption that the number of different keys is upper bounded. We could enforce this upper bound by limiting the size of the `ids` array instead. Eg resetting it when it reaches 1000 items?

i agree with the idea of 1000 keys,  a @Jeroeny  ?

**Comment by Jeroeny** [comment]
> > But for hashing, this would still be useful.
> > This property has been added with the assumption that the number of different keys is upper bounded. We could enforce this upper bound by limiting the size of the `ids` array instead. Eg resetting it when it reaches 1000 items?
> 
> i agree with the idea of 1000 keys, a @Jeroeny ?

Sounds good to me. Upon reaching the limit, do you want to purge the entire array or slice it so that it's maxed out at 1000 items (like the ArrayAdapter cache) ?

**Comment by nicolas-grekas** [comment]
The logic in ArrayAdapter is too involving for this. I'd suggest halving the array instead:
`array_splice($this->ids, 0, 500);`

File: src/Symfony/Component/Cache/Traits/AbstractAdapterTrait.php [hunk 2]
```diff
@@ -364,6 +364,10 @@ private function generateItems(iterable $items, array &$keys): \Generator
 
     private function getId($key)
     {
+        if (count($this->ids) > 1000) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Traits/AbstractAdapterTrait.php]: I'd suggest moving this after this line:
image

Please also add a leading `\` to `\count()`

**Comment by nicolas-grekas** [comment]
Thank you @a1812.

**Pull request merged**